### PR TITLE
[Core] Port GcsPublisher to Cython

### PR DESCRIPTION
**Pull request by pcmoritz**
## Why are these changes needed?

Next step in https://github.com/ray-project/ray/pull/33769

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -1732,6 +1735,65 @@ cdef class GcsClient:
...
+        error_info.set_job_id(job_id.binary())
+        error_info.set_type(error_type)
+        error_info.set_error_message(message)
+        error_info.set_timestamp(time.time())
+
+        with nogil:
+            check_status(
+                self.inner.get().PublishError(c_key_id, error_info, c_num_retries))
+
+    def publish_logs(self, log_json):
```
Comment (rkooo567) [inline python/ray/_raylet.pyx:1766]: Nit; specify the type of log_json? Is it the dictionary?

File: python/ray/_raylet.pyx [hunk 2]
```diff
@@ -1732,6 +1735,65 @@ cdef class GcsClient:
...
+        cdef:
+            CLogBatch log_batch
+            c_string c_job_id
+
+        job_id = log_json.get("job")
+        log_batch.set_ip(log_json.get("ip") if log_json.get("ip") else b"")
+        log_batch.set_pid(str(log_json.get("pid")) if log_json.get("pid") else b"")
```
Comment (rkooo567) [inline python/ray/_raylet.pyx:1773]: OOC, this means pid field will have both string & bytes type? (when pid exists, it is str(pid), and when it doesn't it is `b""`).

File: src/ray/gcs/pubsub/gcs_pub_sub.cc [hunk 3]
```diff
@@ -212,5 +213,87 @@ Status GcsSubscriber::SubscribeAllWorkerFailures(
...
+  RAY_CHECK(address.size() == 2);
+  gcs_address_ = address[0];
+  gcs_port_ = std::stoi(address[1]);
+}
+
+Status PythonGcsPublisher::Connect() {
+  grpc::ChannelArguments arguments;
+  arguments.SetInt(GRPC_ARG_MAX_MESSAGE_LENGTH, 512 * 1024 * 1024);
+  arguments.SetInt(GRPC_ARG_KEEPALIVE_TIME_MS, 60 * 1000);
+  arguments.SetInt(GRPC_ARG_KEEPALIVE_TIMEOUT_MS, 60 * 1000);
```
Comment (rkooo567) [inline src/ray/gcs/pubsub/gcs_pub_sub.cc:228]: Should we use the ray config value for these values? I found we have the same config for other Python client. 

```
Status PythonGcsClient::Connect() {
  grpc::ChannelArguments arguments;
  arguments.SetInt(GRPC_ARG_MAX_MESSAGE_LENGTH, 512 * 1024 * 1024);
  arguments.SetInt(GRPC_ARG_KEEPALIVE_TIME_MS, 60 * 1000);
  arguments.SetInt(GRPC_ARG_KEEPALIVE_TIMEOUT_MS, 60 * 1000);
  channel_ = rpc::BuildChannel(options_.gcs_address_, options_.gcs_port_, arguments);
  kv_stub_ = rpc::InternalKVGcsService::NewStub(channel_);
  runtime_env_stub_ = rpc::RuntimeEnvGcsService::NewStub(channel_);
  node_info_stub_ = rpc::NodeInfoGcsService::NewStub(channel_);
  job_info_stub_ = rpc::JobInfoGcsService::NewStub(channel_);
  return Status::OK();
```

File: src/ray/gcs/pubsub/gcs_pub_sub.h [hunk 4]
```diff
@@ -132,5 +133,26 @@ class GcsSubscriber {
   const std::unique_ptr<pubsub::SubscriberInterface> subscriber_;
 };
 
+// This client is only supposed to be used from Cython / Python
+class RAY_EXPORT PythonGcsPublisher {
+ public:
+  explicit PythonGcsPublisher(const std::string &gcs_address);
+  Status Connect();
```
Comment (rkooo567) [inline src/ray/gcs/pubsub/gcs_pub_sub.h:140]: Can you add docstring for the interface?

**Review by rkooo567 — APPROVED** [review]
LGTM. Nit comments

File: python/ray/_raylet.pyx (hunk 2 above)
Comment (pcmoritz) [reply python/ray/_raylet.pyx]: This is fixed now :)

File: dashboard/dashboard.py [hunk 5]
```diff
@@ -261,7 +260,7 @@ def sigterm_handler():
             raise e
 
         # Something went wrong, so push an error to all drivers.
-        gcs_publisher = GcsPublisher(address=args.gcs_address)
+        gcs_publisher = ray._raylet.GcsPublisher(address=args.gcs_address)
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply dashboard/dashboard.py:263]: I don't know what the new publisher is, but the C++ wrapped code uses the same configs as the old Python code (which I believe uses the same configs as the aio publisher).

The config was taken from https://github.com/ray-project/ray/blob/master/python/ray/_private/gcs_utils.py#L72

File: python/ray/_raylet.pyx (hunk 1 above)
Comment (pcmoritz) [reply python/ray/_raylet.pyx]: Fixed, yes :)

File: src/ray/gcs/pubsub/gcs_pub_sub.h (hunk 4 above)
Comment (pcmoritz) [reply src/ray/gcs/pubsub/gcs_pub_sub.h:145]: fixed

File: src/ray/gcs/pubsub/gcs_pub_sub.cc [hunk 6]
```diff
@@ -236,16 +236,15 @@ constexpr int MAX_GCS_PUBLISH_RETRIES = 60;
...
-  if (timeout_ms != -1) {
-    context.set_deadline(std::chrono::system_clock::now() +
-                         std::chrono::milliseconds(timeout_ms));
-  }
-
   int count = num_retries == -1 ? MAX_GCS_PUBLISH_RETRIES : num_retries;
   rpc::GcsPublishReply reply;
   grpc::Status status;
   while (count > 0) {
+    grpc::ClientContext context;
```
Comment (scv119) [inline src/ray/gcs/pubsub/gcs_pub_sub.cc:243]: i see, so the problem is the context got reused?

Comment (pcmoritz) [reply src/ray/gcs/pubsub/gcs_pub_sub.cc:243]: yes, that's right! @vitsai just told me she knew that and it is actually in the gRPC docs 😄

File: src/ray/gcs/pubsub/gcs_pub_sub.cc (hunk 3 above)
Comment (pcmoritz) [reply src/ray/gcs/pubsub/gcs_pub_sub.cc]: I dedup'ed the code now -- I don't think it is a great idea to introduce additional configs here (they were also hard coded in Python).

Once we are done with this migration, we can consider unifying the synchronous (wrapped by Python) and asynchronous (used by C++) configs :)

**Pull request merged**